src/GraphQL/Utilities/Federation/Proposal.cs

```diff
@@ -0,0 +1,243 @@
+        public static FieldBuilder<TSourceType, TReturnType> Provides<TSourceType, TReturnType>(this FieldBuilder<TSourceType, TReturnType> fieldBuilder, string fields) => BuildAstMeta(fieldBuilder, "provides", fields);
+        public static FieldBuilder<TSourceType, TReturnType> External<TSourceType, TReturnType>(this FieldBuilder<TSourceType, TReturnType> fieldBuilder) => BuildAstMeta(fieldBuilder, "external");
+    }
+
+    /// <summary>
+    /// Can be replaced with ObjectGraphType extensions, the same as above - idea to allow us to define type directives, e.g.:
+    /// public class AcmeType: ObjectGraphType {
+    ///   public AcmeType() {
+    ///     // Metadata["__AST_MetaField__"] = new GraphQLObjectTypeDefinition() {...};
+    ///     Key("id");
+    ///   }
+    /// }
+    ///
+    /// TODO:
+    /// - there is only two cases: `type Account @key(fields:"id")` and `extend type Account @key(fields:"id")`
+    /// - will be nice to have to enforce declaration one of above
+    /// - will be nice to have to enforce resolve reference declaration
+    /// </summary>
+    /// <typeparam name="TSourceType"></typeparam>
+    public class FederatedObjectGraphType<TSourceType> : ObjectGraphType<TSourceType>
+    {
+        public void Key(string fields) => this.BuildAstMeta("key", fields);
+
+        public void ExtendByKey(string fields)
+        {
+            this.BuildExtensionAstMeta("key", fields);
+            Key(fields);
+        }
+
+        // Resolve reference is from schema first aproach
+        public void ResolveReferenceAsync(Func<FederatedResolveContext, Task<TSourceType>> resolver) => ResolveReferenceAsync(new FuncFederatedResolver<TSourceType>(resolver));
+
+        public void ResolveReferenceAsync(IFederatedResolver resolver)
+        {
+            // Metadata[FederatedSchemaBuilder.RESOLVER_METADATA_FIELD] = resolver;
+            Metadata["__FedResolver__"] = resolver;
+        }
+    }
+
+    // TODO: move me into separate project
+    //public static class FederatedGraphQLBuilderExtensions
+    //{
+    //    public static IGraphQLBuilder AddFederation(this IGraphQLBuilder builder)
+    //    {
+    //        builder.Services.AddSingleton<AnyScalarGraphType>();
+    //        builder.Services.AddSingleton<ServiceGraphType>();
+
+    //        return builder;
+    //    }
+    //}
+
+
+    /// <summary>
+    /// Small extensions just to hide registration of required types, can be extension instead
+    /// </summary>
+    public class FederatedSchemaProposal : Schema
+    {
+        public FederatedSchemaProposal() : this(new DefaultServiceProvider()) { }
+        public FederatedSchemaProposal(IServiceProvider serviceProvider) : base(serviceProvider)
+        {
+            RegisterValueConverter(new AnyValueConverter());
+            RegisterType<AnyScalarGraphType>();
+            RegisterType<ServiceGraphType>();
+            RegisterType<EntityType>();
```

Comment:
do we need a any value converter? I tried without one and it still did work :| I tried that because I couldn't find source for any value converter I also couldn't find it on schema first (same as RegisterValueConverter)

Reply:
also if we were to put this directly on Schema instead of having federation on a separate class, would that cause any issues?